cdap-unit-test/src/test/java/co/cask/cdap/service/FileUploadApp.java

```diff
@@ -0,0 +1,245 @@
+              responder.sendStatus(HttpURLConnection.HTTP_BAD_REQUEST);
+            } else {
+              responder.sendStatus(HttpURLConnection.HTTP_INTERNAL_ERROR);
+            }
+          }
+        }
+      };
+    }
+
+    /**
+     * This method functionality is the same as the
+     * {@link #upload(HttpServiceRequest, HttpServiceResponder, String, long)},
+     * except that it returns a {@link HttpContentConsumer} instead of {@link NonTransactionalHttpContentConsumer}
+     * so that it updates the tracking table directly in the onReceive method
+     */
+    @POST
+    @Path("/txUpload/{dataset}/{partition}")
+    public HttpContentConsumer txUpload(HttpServiceRequest request,
+                                        HttpServiceResponder responder,
+                                        @PathParam("dataset") String dataset,
+                                        @PathParam("partition") long partition) throws Exception {
+      final String md5 = request.getHeader("Content-MD5");
+      if (md5 == null) {
+        responder.sendError(HttpURLConnection.HTTP_BAD_REQUEST, "Missing header \"Content-MD5\"");
+        return null;
+      }
+
+      // Construct the partition and the partition location
+      PartitionKey partitionKey = PartitionKey.builder().addLongField("time", partition).build();
+      PartitionedFileSet pfs = getContext().getDataset(dataset);
+      final PartitionOutput partitionOutput = pfs.getPartitionOutput(
+        partitionKey);
+      final Location partitionDir = partitionOutput.getLocation();
+      if (!partitionDir.mkdirs()) {
+        responder.sendError(HttpURLConnection.HTTP_CONFLICT,
+                            String.format("Partition for key '%s' already exists for dataset '%s'",
+                                          partitionKey, dataset));
+        return null;
+      }
+
+      final MessageDigest messageDigest = MessageDigest.getInstance("MD5");
+
+      final Location location = partitionDir.append("upload-" + System.currentTimeMillis());
+      final WritableByteChannel channel = Channels.newChannel(location.getOutputStream());
+
+      // Handle upload content. The onReceived method is transactional.
+      return new HttpContentConsumer() {
+
+        @Override
+        public void onReceived(final ByteBuffer chunk) throws Exception {
+          trackingTable.increment(Bytes.toBytes(chunk.remaining()), 1L);
+          chunk.mark();
+          messageDigest.update(chunk);
+          chunk.reset();
+          channel.write(chunk);
+        }
+
+        @Override
+        public void onFinish(HttpServiceResponder responder) throws Exception {
+          channel.close();
+          String uploadedMd5 = BaseEncoding.base64().encode(messageDigest.digest());
+          if (!md5.equals(uploadedMd5)) {
+            throw new IllegalArgumentException("MD5 not match. Expected '" + md5 + "', received '" + uploadedMd5 + "'");
```

Comment:
just wondering - if this exception is thrown, does the `onError` callback below get triggered?